DIPs/1NNN-MHH.md

```diff
@@ -0,0 +1,242 @@
+### A less simple example
+There are, however, situations where a UDA is highly useful but there does not exist as clean of a hierarchy as in the previous example.
+
+For example, an attribute will often be attached to a - in effect - free standing symbol like a `unittest`. 
+```D
+enum runMyTest;
+
+@runMyTest unittest {
+
+}
+```
+Since most of these declarations are at or near module scope, we can simply iterate over *all* the `unittest`s looking for one's matching our UDA. Although not technically required, it is still common to find these tests in a pre-compilation step - or by manually setting the test runner to work using a `mixin` statement.
+
+This pattern is fine for `unittest`s: When used, there are usually many of them, and when there are not many the potential for a performance hit is mitigated by the compiler providing a list of `unittest`s within a given scope.
+
+### The behaviour to be enabled by this DIP.
+There are, however, situations where this pattern is less than ideal. Some UDAs may be used to declare things which are both sparse and not `unittest`s. 
+
+If we setup a mechanism to get these UDA-ed symbols - assuming we know where they are i.e. a recursive search is currently not possible but scopes containing UDAs can be declared using a different UDA - we still have to iterate over *every* declaration in a given scope looking for those with our desired set of attributes. For a sparse declaration - that is, if we are looking for a few declarations out of a several thousand line file - this is not an efficient way of doing things.
+
+For example, this DIP is motivated specifically by a desire to be able to declare benchmarks in the following manner (without needing the user to `mixin` anything):
+
+```D
+@Benchmark!(SomeInformationAboutTheBenchmark)
+float dotp(float[3] vec1, float[3] vec2) pure
+{
+    return vec1[0] * vec2[0] + vec1[1] * vec2[1] + vec1[2] * vec2[2] 
+}
+//The library can then do what it wants with dotp, be that schedule it to run or add it to a table at compile time
+```
+
+This DIP proposes a simple solution to this problem: Let the UDA see sideways, that is, let it see what it is attached to. Rather than being some paradigm shift in how UDAs are used, this should simply be viewed as having the same effect of adding syntactic sugar for the following construct:
+```D
+    module home;
+    enum MyUDA;
+    template HandleTheUDA(alias handleThis) {/* impl */}
+    @MyUDA void widget() {}
+    //Ugly
+    mixin HandleTheUDA!handleThis;
+    //Slow, we don't want to search the entire module just for one UDA
+    mixin HandleTheUDA!home;
+```
+i.e. With this DIP, MyUDA can be declared as a template that can perform the `mixin` itself. This does not add any new dragons or side effects beyond what the original construct could do.
```

Comment:
In your examples, show how the situation is *with* your changes. Only showing the problem case and vaguely explaining what changes is a little confusing. Reading this, I see «`MyUDA` can be declared as a template» and wonder what you're talking about; `MyUDA` clearly is not a template in your example. It would be fine if you showed the changes after that paragraph.

Reply:
"This DIP enables the library writer to create a UDA that can manage the per-symbol `mixin` itself. In translating our example to this new idiom *MyUDA* would go from being an `enum` to a template as the code contained within *HandleTheUDA* is now done at the UDA instead. When the library is consumed, however, the API would be the same due to the use of default parameters (other than the lack of `mixin`s)."

Better?